bitcoin/BRSyncManager.c

```diff
@@ -0,0 +1,1549 @@
+    }
+    assert (0);
+    return NULL;
+}
+
+static void
+BRPeerSyncManagerFree(BRPeerSyncManager manager) {
+    BRPeerManagerDisconnect (manager->peerManager);
+
+    pthread_mutex_destroy(&manager->lock);
+    memset (manager, 0, sizeof(*manager));
+    free (manager);
+}
+
+static void
+BRPeerSyncManagerConnect(BRPeerSyncManager manager) {
+    BRPeerManagerConnect (manager->peerManager);
+    int8_t connected = BRPeerStatusDisconnected != BRPeerManagerConnectStatus (manager->peerManager);
+    if (0 == pthread_mutex_lock (&manager->lock)) {
+        if (connected && !manager->isConnected) {
+            manager->isConnected = 1;
+        } else {
+            connected = 0;
+        }
+        pthread_mutex_unlock (&manager->lock);
+    } else {
+        assert (0);
+    }
+
+    // send event on transition to connected state
+    if (connected) {
+        manager->eventCallback (manager->eventContext,
+                                BRPeerSyncManagerAsSyncManager (manager),
+                                (BRSyncManagerEvent) {
+                                    SYNC_MANAGER_CONNECTED,
+                                });
+    };
+}
+
+static void
+BRPeerSyncManagerDisconnect(BRPeerSyncManager manager) {
+    BRPeerManagerDisconnect (manager->peerManager);
+    int8_t disconnected = BRPeerStatusDisconnected == BRPeerManagerConnectStatus (manager->peerManager);
+    if (0 == pthread_mutex_lock (&manager->lock)) {
+        if (disconnected && manager->isConnected) {
+            manager->isConnected = 0;
+        } else {
+            disconnected = 0;
+        }
+        pthread_mutex_unlock (&manager->lock);
+    } else {
+        assert (0);
+    }
+
+    // send event on transition to disconnected state
+    if (disconnected) {
+        manager->eventCallback (manager->eventContext,
+                                BRPeerSyncManagerAsSyncManager (manager),
+                                (BRSyncManagerEvent) {
+                                    SYNC_MANAGER_DISCONNECTED,
+                                });
+    };
+}
+
+static void
+BRPeerSyncManagerScan(BRPeerSyncManager manager) {
+    BRPeerManagerRescan (manager->peerManager);
+}
+
+typedef struct {
+    BRPeerSyncManager manager;
+    BRTransaction *transaction;
+} SubmitTransactionInfo;
+
+static void
+BRPeerSyncManagerSubmit(BRPeerSyncManager manager,
+                        OwnershipGiven BRTransaction *transaction) {
+    SubmitTransactionInfo *info = malloc (sizeof (SubmitTransactionInfo));
+    info->manager = manager;
+    info->transaction = transaction;
+
+    BRPeerManagerPublishTx (manager->peerManager,
+                            transaction,
+                            info,
+                            _BRPeerSyncManagerTxPublished);
+}
+
+static void
+BRPeerSyncManagerTickTock(BRPeerSyncManager manager) {
+    double progress = BRPeerManagerSyncProgress (manager->peerManager, 0);
+    uint8_t needSyncEvent = progress < 1.0;
```

Comment:
`0.0 < progress && progress < 1.0`?  Symmetrically.

Reply:
I had that, took it out, thought they might want the 0% case announced. Can put back in.